### ISSUE TITLE
'xmlhttprequest' instrumentation breaking transactions early

### ISSUE THREAD

**Issue by Chuli2k** [elastic/apm-agent-rum-js#390]
In my angular app I noticed that the "xmlhttprequest" instrumentation was ending my transactions too early. Which means that some of my http calls did not go into the APM server.
I was playing around with the code and found that if I add a task at the start, then all the http calls get rgistered correctly.

Here is the relevant code. If you comment out the 2 lines then everything is in the transaction. If not then the last 2 http calls are lost.
```ts
async ngOnInit() {
    const tr = this.apm.apm.startTransaction('test', 'custom');
    // const task1 = tr.addTask();
    const r1 = await this.http.get('http://localhost:52516/api/Test/Ping').toPromise(); // return OK
    const r2 = await this.http.get('http://localhost:52516/api/Test/Ping2').toPromise(); // wait 500ms then return OK
    const r3 = await this.http.get('http://localhost:52516/api/Test/Ping2').toPromise(); // wait 500ms then return OK
    // tr.removeTask(task1);
    tr.end();
  }
```
Is this intended behaviour? Becouse if it is, the lack of task documentation is really annoying... First I was trying to solve the issue with a custom span, but because it doesn't create a new task it did nothing.

And also the console log:
```
core.js:16829 Angular is running in the development mode. Call enableProdMode() to enable the production mode.
logging-service.js:76 TransactionService.startTransaction Transaction {options: {…}, name: "test", type: "custom", id: "3b18bed481132ce8", traceId: "44203c0c3f627dcbb2ea833fc91571ee", …}
logging-service.js:76 TransactionService.startSpan GET http://localhost:52516/api/Test/Ping external.http
logging-service.js:76 TransactionService.addTask task1
logging-service.js:76 TransactionService.startSpan GET http://localhost:4200/sockjs-node/info external.http
logging-service.js:76 TransactionService.addTask task2
logging-service.js:76 TransactionService.removeTask task2
logging-service.js:76 TransactionService.removeTask task1
logging-service.js:76 TransactionService transaction finished Transaction {options: {…}, name: "test", type: "custom", id: "3b18bed481132ce8", traceId: "44203c0c3f627dcbb2ea833fc91571ee", …}
logging-service.js:76 TransactionService.add Transaction {options: {…}, name: "test", type: "custom", id: "3b18bed481132ce8", traceId: "44203c0c3f627dcbb2ea833fc91571ee", …}
logging-service.js:76 TransactionService.startSpan GET http://localhost:52516/api/Test/Ping2 external.http
logging-service.js:76 TransactionService.addTask task1
logging-service.js:76 TransactionService.removeTask task1
logging-service.js:76 TransactionService.startSpan GET http://localhost:52516/api/Test/Ping2 external.http
logging-service.js:76 TransactionService.addTask task1
logging-service.js:76 TransactionService.removeTask task1
```

Example project: https://github.com/Chuli2k/ang-rum-routing

**Comment by vigneshshanmugam** [comment]
**TLDR**

+ XHR tasks are scheduled as macroTasks on Angular applications since zone.js schedules them in the macrotask queue. 
+ The RUM agent XHR tasks always gets scheduled/invoked before zone.js invokes them and as a result the transaction is ended even before all the XHR calls are scheduled. 
+ This issue only affects users of angular and zone.js. 

The easier fix would be schedule the XHR calls as macrotasks via `setTimeout` or `setImmediate`. You can see how the calls are scheduled with our debug mode enabled. 
```
TransactionService.startTransaction Transaction
 TransactionService.startSpan GET http://localhost:8003/data external.http
 TransactionService.addTask task1
[Zone.js] Invoking zone js task as macrotask
 TransactionService.startSpan GET http://localhost:8003/data external.http
 TransactionService.addTask task2
[Zone.js] Invoking zone js task as macrotask
 TransactionService.startSpan GET http://localhost:8003/data external.http
 TransactionService.addTask task3
 TransactionService.removeTask task1
[Zone.js] Invoking zone js task as macrotask
 TransactionService transaction finished Transaction ```

**Comment by jahtalab** [comment]
Digging a bit deeper into the issue, It seems that the zone.js checks if any `load` event listeners are registered on the xhr and if there are any, it waits for all of them before invoking the macroTask which results in the promise being resolved! However the RUM agent only registers a`readystatechange` event listener. 
Since the order of execution of event listeners is not guaranteed this might be potentially an issue for other applications as well.

The solution for would to instrument both `load`, `readystatechange` events and wait for listeners to be executed first before invoking the task!
